modules/standard/Heap.chpl

```diff
@@ -0,0 +1,400 @@
+      _enter();
+      var result = _data.isEmpty();
+      _leave();
+      return result;
+    }
+
+    /*
+      Return the maximal element in the heap.
+
+      :return: The maximal element in the heap
+      :rtype: `eltType`
+
+      .. note::
+        *Maximal* is defined by ``comparator``. If a ``reverseComparator`` is passed to ``init``,
+        the heap will return the minimal element.
+
+    */
+    proc top() {
+      if (isOwnedClass(eltType)) {
+        compilerError("top() method is not avaliable on a 'heap'",
+                      " with elements of a owned type, here: ",
+                      eltType: string);
+      }
+      _enter();
+      if (boundsChecking && isEmpty()) {
+        boundsCheckHalt("Called \"heap.top\" on an empty heap.");
+      }
+      var result = _data[0];
+      _leave();
+      return result;
+    }
+
+    /*
+      Wrapper of comparing elements
+    */
+    pragma "no doc"
+    proc _greater(x: eltType, y: eltType) {
+      return chpl_compare(x, y, comparator) > 0;
+    }
+
+    /*
+      helper procs to maintain the heap
+    */
+    pragma "no doc"
+    proc _heapify_up(in pos: int) {
+      while (pos) {
+        var parent = pos / 2;
+        if (_greater(_data[pos],_data[parent])) {
+          _data[parent] <=> _data[pos];
+          pos = parent;
+        }
+        else break;
+      }
+    }
+
+    pragma "no doc"
+    proc _heapify_down(in pos: int) {
+      while (pos < _data.size) {
+        // find the child node with greater value
+        var greaterChild = pos*2;
+        if (greaterChild >= _data.size) then break; // reach leaf node, break
+        if (greaterChild + 1 < _data.size) {
+          // if the right child node exists
+          if (_greater(_data[greaterChild+1],_data[greaterChild])) {
+            // if the right child is greater, then update the greaterChild
+            greaterChild += 1;
+          }
+        }
+        // if the greaterChild's value is greater than current node, then swap and continue
+        if (_greater(_data[greaterChild],_data[pos])) {
+          _data[greaterChild] <=> _data[pos];
+          pos = greaterChild;
+        }
+        else break;
+      }
+    }
+
+    /*
+      Push an element into the heap
+
+      :arg element: The element that will be pushed
+      :type element: `eltType`
+    */
+    proc push(in element: eltType)
+    lifetime this < element {
+      _enter();
+      _data.append(element);
+      _heapify_up(_data.size-1);
+      _leave();
+    }
+
+    /*
+      Push elements of a list into the heap
+
+      :arg x: The list of which elements is to push.
+      :type x: `list(eltType)`
+    */
+    proc push(ref x: list(eltType)) {
+      for e in x do
+        push(e);
+    }
+
+    /*
+      Push elements in an array into a heap.
+
+      :arg x: The array of which elements is to push.
+      :type x: `[?d] eltType`
+    */
+    proc push(ref x:[?d] eltType) {
+      for e in x do
+        push(e);
+    }
```

Comment:
I think no. Let M be the size of the array and N be the size of the heap. The operation is O(MlgN). Another way to do this is to reconstruct the whole heap, which takes O(N+M). Well, O(N+M) actually grows slower than O(MlgN) but I don't think that's good...

Reply:
I think there can be something we leverage still, but by no means it should happen now. But can you add a comment saying that we can do some optimizations there if needed?

At the very least, we can have a low-level implementation to avoid locking and unlocking for every element and lock and unlock only once. In any case, if we have an overloads such as these, we should always think about potential performance benefits.

Reply:
Generally speaking, when M is very large and N is relatively small. It's faster to construct a new heap. In the other case, it's faster to insert element one by one. But I don't know how to give a precise bound to determine when to use which. After all, the time complexity is a kind of estimation.

Reply:
https://github.com/chapel-lang/chapel/pull/15997/commits/4455d14c258724908932a1b4813401337ce88e75 Lock/unlock once when pushing